frontend/cypress/integration/common/app_details.ts

```diff
@@ -0,0 +1,71 @@
+import { Then } from 'cypress-cucumber-preprocessor/steps';
+
+const APP = 'details';
+const NAMESPACE = 'bookinfo';
+
+Then('user sees details information for app', () => {
+  cy.getBySel('app-description-card').within(() => {
+    cy.get('#pfbadge-A').parent().parent().contains('details'); // App
+    cy.get('#pfbadge-W').parent().parent().contains('details-v1'); // Workload
+    cy.get('#pfbadge-S').parent().parent().contains('details'); // Service
+  });
+});
+
+Then('user sees a minigraph', () => {
+  cy.getBySel('mini-graph');
+});
+
+function openTab(tab: string) {
+  cy.get('#basic-tabs').should('be.visible').contains(tab).click();
+}
+
+Then('user sees inbound and outbound traffic information', () => {
+  openTab('Traffic');
+  cy.contains('Inbound Traffic');
+  // Details app should have some inbound traffic.
+  cy.contains('No Inbound Traffic').should('not.exist');
+  cy.contains('Outbound Traffic');
+});
+
+Then('user sees inbound metrics information', () => {
+  cy.intercept(Cypress.config('baseUrl') + `/api/namespaces/${NAMESPACE}/apps/${APP}/dashboard*`).as('fetchMetrics');
+  openTab('Inbound Metrics');
+  cy.wait('@fetchMetrics');
+  cy.waitForReact(1000, '#root');
+  cy.getReact('IstioMetrics', { props: { 'data-test': 'inbound-metrics-component' } })
+    // Theres be multiple components of the same name providing state through redux/hooks.
+    // This gets us the one we we want but this tests will fail if the context providers
+    // change names or are removed.
+    .nthNode(2)
```

Comment:
Just curious, why the second node?.

Reply:
There's two parent components from the router and from redux that both match the `cy.getReact('IstioMetrics')` query. We want the actual component that we added which is the bottom one at index 2. Wish there was a way to specify the "last" one but doesn't look like it.

![Screenshot from 2022-05-13 14-12-52](https://user-images.githubusercontent.com/6226732/168343002-da77afca-f0ee-423d-b23d-c9384cbe48e1.png)

Reply:
I see, thanks.